### PR TITLE
Updated SIO to work with the latest stdexec release (as of 08/13), as well as a few other updates

### PR DESCRIPTION
**Pull request by emmett2020**
Changes:
1. Add the configuration file for [clangd](https://clangd.llvm.org/). clangd is a language server.
2. Using FetchContent to download stdexec and catch2. 
3. Refine CI since we could automatically download needed third parities. 
4. Upgrade catch2 from version 2 to version 3 to speed up compilation.
5. Remove several tags in concepts.
6. Update almost all senders, add customized domain, well-formed early check, basic_sequence_sender, etc.
7. Some of the sender implementations have changed. For example, the dynamic_operation for merge_each is temporarily removed and can be added back later.
8. Add more test cases.
9. Update date of copyright

Results:
All tests passed and all examples worked as expected.

### CODE REVIEW

File: .clangd [hunk 1]
```diff
@@ -0,0 +1,4 @@
+# https://clangd.llvm.org/config
+
+CompileFlags:
+  Remove: -fconcepts-diagnostics-depth=10
```
Comment (maikel) [inline .clangd:4]: why is this needed? What is the effect?

File: .github/workflows/cmake.yml [hunk 2]
```diff
@@ -18,34 +18,7 @@ jobs:
...
-      run: cmake -B ${{github.workspace}}/catch2-build -S ${{github.workspace}}/catch2 -DCMAKE_INSTALL_PREFIX=${{github.workspace}}/catch2-install -DBUILD_TESTING=OFF
-
-    - name: Build Catch2
-      run: cmake --build ${{github.workspace}}/catch2-build --parallel 4
-
-    - name: Install Catch2
-      run: cmake --install ${{github.workspace}}/catch2-build
-
-    - uses: actions/checkout@v3
+    - uses: actions/checkout@v4
```
Comment (maikel) [inline .github/workflows/cmake.yml:21]: awesome

File: cmake/FetchCatch2.cmake [hunk 3]
```diff
@@ -0,0 +1,8 @@
+include(FetchContent)
+FetchContent_Declare(
+  Catch2
+  GIT_REPOSITORY https://github.com/catchorg/Catch2.git
+  GIT_TAG v3.6.0
+)
+message(STATUS "Downloading Catch2......")
```
Comment (maikel) [inline cmake/FetchCatch2.cmake]:
```suggestion
message(STATUS "Downloading Catch2")
```

File: cmake/FetchStdexec.cmake [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+FetchContent_Declare(
+  stdexec
+  GIT_REPOSITORY https://github.com/NVIDIA/stdexec.git
+
+  # branch: main. date: 2024-08-13
+  GIT_TAG 54b38c99e8cbb348c1dd3508fa540728882247db
+)
+
+set(STDEXEC_BUILD_EXAMPLES OFF CACHE BOOL "close stdexec examples")
+message(STATUS "Downloading stdexec......")
```
Comment (maikel) [inline cmake/FetchStdexec.cmake]:
```suggestion
message(STATUS "Downloading stdexec")
```

File: source/sio/ip/resolve.hpp [hunk 5]
```diff
@@ -340,8 +340,8 @@ namespace sio::async {
       ip::resolver_query query_;
 
       template <decays_to<sender> Self, class Receiver>
-      static operation<Scheduler, Receiver>
-        subscribe(Self&& self, exec::subscribe_t, Receiver&& receiver) {
+      friend operation<Scheduler, Receiver>
+        tag_invoke(exec::subscribe_t, Self&& self, Receiver receiver) {
```
Comment (maikel) [inline source/sio/ip/resolve.hpp:344]: changing the sequence related cpos to member cpos need changes in stdexec, right?

**Review by maikel — COMMENTED** [review]
This is awesome. Please add a copyright notice with your name everywhere where you want to.

I will play with this a bit and merge it soon. Do you have anything in mind that needs further improvements?

File: .github/workflows/cmake.yml (hunk 2 above)
Comment (maikel) [reply .github/workflows/cmake.yml:21]: we don't need to pass stdexec_DIR and catch2_dir to cmake if we use fetchcontent, right?

File: .clangd (hunk 1 above)
Comment (emmett2020) [reply .clangd:4]: Yes. About "**why this is needed**":  some guys would like to use clangd, a popular Language Protocol Server provided by llvm, to do "code completion", "go to definition" and so on. So I added this config file.
    About "**what is the effect**":  The "-fconcepts-diagnostics-depth" option is provided by gcc not by clang compiler. However, clangd is based on clang compiler. So I removed this flag to suppress clangd's warning.
    More details: https://clangd.llvm.org/

File: .github/workflows/cmake.yml (hunk 2 above)
Comment (emmett2020) [reply .github/workflows/cmake.yml:21]: Yes, we don't need to pass stdexec_DIR and catch2_DIR to cmake since "fetchcontent"  can automatically download these libraries in first time compilation.

File: source/sio/ip/resolve.hpp (hunk 5 above)
Comment (emmett2020) [reply source/sio/ip/resolve.hpp:344]: Sure, we should adjust stdexec first.

File: cmake/FetchStdexec.cmake (hunk 4 above)
Comment (emmett2020) [reply cmake/FetchStdexec.cmake]: Thx, fixed this.

File: cmake/FetchCatch2.cmake (hunk 3 above)
Comment (emmett2020) [reply cmake/FetchCatch2.cmake]: Good point, I have fixed this. Thanks.

**Comment by emmett2020** [comment]
> This is awesome. Please add a copyright notice with your name everywhere where you want to.
> 
> I will play with this a bit and merge it soon. Do you have anything in mind that needs further improvements?

Thanks for your time👍 I've added my new name to copyright. I think I don't need new content in this MR, unless there is a new review comment   :)
